src/psfmachine/machine.py

```diff
@@ -837,289 +635,197 @@ def build_time_model(self, plot=False, downsample=False, split_time_model=False)
-                "pos_corr",
-                "centroid",
-            ]:
-                A3 = _combine_A(
-                    A2, poscorr=[poscorr1_binned[seg_mask], poscorr2_binned[seg_mask]]
-                )
-            else:
-                A3 = _combine_A(A2, time=time_binned[seg_mask])
-
-            # No saturated pixels, 1e5 is a hardcoded value for Kepler.
-            k = (
-                (flux_binned_raw < 1e5).all(axis=0)[None, :]
-                * np.ones(flux_binned_raw[seg_mask].shape, bool)
-            ).ravel()
-            # No faint pixels, 100 is a hardcoded value for Kepler.
-            k &= (
-                (flux_binned_raw[seg_mask] > 100).all(axis=0)[None, :]
-                * np.ones(flux_binned_raw[seg_mask].shape, bool)
-            ).ravel()
-            # No huge variability
-            k &= (
-                (np.abs(flux_binned[seg_mask] - 1) < 1).all(axis=0)[None, :]
-                * np.ones(flux_binned[seg_mask].shape, bool)
-            ).ravel()
-            # No nans
-            k &= np.isfinite(flux_binned[seg_mask].ravel()) & np.isfinite(
-                flux_err_binned[seg_mask].ravel()
+            # normalize components
+            mpc1_smooth = (mpc1_smooth - mpc1_smooth.mean()) / (
+                mpc1_smooth.max() - mpc1_smooth.mean()
             )
+            mpc2_smooth = (mpc2_smooth - mpc2_smooth.mean()) / (
+                mpc2_smooth.max() - mpc2_smooth.mean()
+            )
+            # combine them as the first order
+            other_vectors = [mpc1_smooth, mpc2_smooth, mpc1_smooth * mpc2_smooth]
+
+        # create a 3D perturbation matrix
+        P3 = PerturbationMatrix3D(
+            time=_whitened_time,
+            dx=dx,
+            dy=dy,
+            segments=split_time_segments,
+            focus=focus_component,
+            resolution=self.n_time_points,
+            other_vectors=other_vectors,
+            bin_method=bin_method,
+        )
 
-            # we solve the segment by using `seg_mask` on all `*_binned` variables
-            for count in [0, 1, 2]:
-                sigma_w_inv = A3[k].T.dot(
-                    (
-                        A3.multiply(1 / flux_err_binned[seg_mask].ravel()[:, None] ** 2)
-                    ).tocsr()[k]
-                )
-                sigma_w_inv += np.diag(1 / prior_sigma ** 2)
-                # Fit the flux - 1
-                B = A3[k].T.dot(
-                    (
-                        (flux_binned[seg_mask].ravel() - 1)
-                        / flux_err_binned[seg_mask].ravel() ** 2
-                    )[k]
-                )
-                B += prior_mu / (prior_sigma ** 2)
-                time_model_w = np.linalg.solve(sigma_w_inv, B)
-                res = flux_binned[seg_mask] - A3.dot(time_model_w).reshape(
-                    flux_binned[seg_mask].shape
+        # get uncontaminated pixel norm-flux
+        flux, flux_err = np.array(
```

Comment:
This should be the uncontaminated flux, not source flux

Reply:
I see what you mean, but we're using `flux, flux_err` to mask good-behaved pixels to then use that mask `k` in `P3` which was created with `source_mask` number of pixels instead of uncontaminated pixels.